Real v5 cel shading

sauraen opened the pull request:
Cel shading has been translated to fast-64/V5 (formatting only) and then to thecozies/V5 (actual V5).

Lighten/darken mode has been removed and replaced with "fixed" mode, where a tint level and tint color is set in the DL, or "segment" mode, where a segment contains (possibly) multiple tint settings like:
```
gsDPSetPrimColor(0, 0, lightTintColorR, lightTintColorG, lightTintColorB, lightTintLevel),
gsSPEndDisplayList(),
gsDPSetPrimColor(0, 0, darkTintColorR, darkTintColorG, darkTintColorB, darkTintLevel),
gsSPEndDisplayList(),
```
enabling the tint colors and levels to be set dynamically from actor code to reflect nearby lighting conditions.

There is no previewing of any of this; what is shown in the GUI is the material without any tint. It will take noticeable changes to the nodes to support previewing this.

sauraen committed:
First pass translating cel shading code formatting

sauraen committed:
Formatting tweaks

sauraen committed:
Draft of GUI for OoT scene lighting

sauraen committed:
Time of Day emulation working

sauraen committed:
Added listening to changes in OoT lighting scene properties and tweaked GUI

sauraen committed:
Whitespace tweaks

sauraen committed:
Merged formatting V5 changes into real V5 branch

sauraen committed:
Fixed vtxList to vertexList

sauraen committed:
In progress changing from lighten darken to tint or segment

sauraen committed:
Changed from lighten-darken to fixed or segmented tint value

sauraen committed:
Tweaks to GUI

thecozies commented on fast64_internal/oot/oot_f3d_writer.py:
```diff
@@ -160,6 +160,8 @@ def ootConvertMeshToC(originalObj, finalTransform, f3dType, isHWv1, name, folder
 		data.source += '#include "' + folderName + '.h"\n\n'
 	else:
 		data.source += '\n'
+	if bpy.context.scene.celShadingPatch:
+		data.source += '#define G_CELSHADING 0x00400000\n\n'
```
the microcode patch should probably have this defined in `gbi.h`, or else there could be a risk of getting a bunch of redefinition errors

sauraen replied:
`gbi.h` is provided by z64hdr in the case of Triforce%, so it's not possible to be changed in that project. Perhaps in a decomp-based toolchain people will be willing to edit their own `gbi.h` when they apply the microcode change as well. However, please keep in mind that redefining a macro to the exact same thing as it was already defined to is not an error, it's silently ignored by the preprocessor.

thecozies commented on fast64_internal/f3d/f3d_writer.py:
```diff
@@ -874,10 +875,10 @@ def __init__(
         currentGroupIndex,
-        triList,
-        vtxList,
+        triGroup,
```
```suggestion
        triGroup: FTriGroup,
```
when changing params here i'd like to start adding type annotations

thecozies commented on fast64_internal/f3d/f3d_writer.py:
```diff
@@ -1613,6 +1700,25 @@ def saveOrGetF3DMaterial(material, fModel, obj, drawLayer, convertTextureData):
 
     return fMaterial, texDimensions
 
+def getCelShadingInfo(material):
```
to me it seems like `CelShadingInfo` is used enough where it could warrant some type of class or definition, as well as the levels. then we can add type annotations everywhere and keep things in sync easier

sauraen replied:
Agreed, I'll add that.

thecozies commented on fast64_internal/f3d/f3d_writer.py:
```diff
@@ -874,10 +875,10 @@ def __init__(
+        triGroup,
         existingVertexData,
         existingVertexMaterialRegions,
+        celShadingInfo
```
if you create a `CelShadingInfo` class this would be a great place for a type annotation

thecozies reviewed (commented):
looks pretty good so far, i have a few suggestions for documentation but i think the implementation should work well. i'll need to test sm64 exports with this and regression test. one concern of mine is making sure that the user's presets are retained

sauraen committed:
Update fast64_internal/render_settings.py

Co-authored-by: thecozies <79979276+thecozies@users.noreply.github.com>

sauraen committed:
Update fast64_internal/render_settings.py

Co-authored-by: thecozies <79979276+thecozies@users.noreply.github.com>

sauraen committed:
Update fast64_internal/render_settings.py

Co-authored-by: thecozies <79979276+thecozies@users.noreply.github.com>

sauraen committed:
Update fast64_internal/f3d/f3d_writer.py

Co-authored-by: thecozies <79979276+thecozies@users.noreply.github.com>

kurethedead committed:
Fix incorrect CI textures on import for OOT (#124)

* Fixed incorrect CI textures on import for OOT.

Fixed bug where CI textures would be corrupted when importing OOT display lists.

* Minor name changes/tweaks.

* Added more informative comments, formatted with black

thecozies committed:
Merge pull request #128 from thecozies/V5

V5

sauraen committed:
Set OoT scene object on scene creation operator

sauraen committed:
Merge branch 'V5' into OoTSceneLighting

sauraen committed:
Merge branch 'V5' into RealV5CelShading

kurethedead committed:
Fixed color attribute add operator data_type argument for Blender 3.3 (#135)

* Fixed operator data_type argument for blender 3.3

* Switched to API instead of using operators

kurethedead committed:
Black formatting (#137)

* Black format files touched by texture saving PR

* Black format files touched by flipbook PR

sauraen committed:
Merge pull request #129 from sauraen/OoTSceneLighting

OoT scene lighting

sauraen committed:
Merged recent V5
